### PR TITLE
zip deployment

### PR DESCRIPTION
**Pull request by gavinfish**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -24,7 +24,7 @@
         <jackson.version>2.7.4</jackson.version>
         <guava.version>19.0</guava.version>
         <powermock.version>1.7.0RC2</powermock.version>
-        <azuresdk.version>1.10.0</azuresdk.version>
+        <azuresdk.version>1.13.0</azuresdk.version>
```
Comment (yungezz) [inline pom.xml]: does this need update in azure-common plugin?

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    @Override
+    public void execute(IWarDeployCommandData context) {
+        WebApp app = context.getWebApp();
+        String deployType = context.getDeployType();
+
+        FilePath srcDir = context.getJobContext().getWorkspace().child(Util.fixNull(context.getSourceDirectory()));
+        try {
+            FilePath[] files = srcDir.list(Util.fixNull(context.getFilePath()));
+
+            if (deployType.equals(Constants.ZIP_DEPLOY)) {
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: switch case?

File: src/main/java/com/microsoft/jenkins/appservice/util/Constants.java [hunk 3]
```diff
@@ -20,6 +20,12 @@ private Constants() {
     // the first option for select element. Keep the same value as jenkins pre-defined default empty value.
     public static final String EMPTY_SELECTION = "- none -";
 
+    public static final String ZIP_FILE_EXTENSION = "zip";
+    public static final String WAR_FILE_EXTENSION = "war";
+    public static final String ZIP_DEPLOY = ZIP_FILE_EXTENSION;
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/util/Constants.java]: enum?

File: src/main/resources/com/microsoft/jenkins/appservice/WebAppDeploymentRecorder/help-deployType.html [hunk 4]
```diff
@@ -0,0 +1,23 @@
+<!--
+  ~ Copyright (c) Microsoft Corporation. All rights reserved.
+  ~ Licensed under the MIT License. See License.txt in the project root for
+  ~ license information.
+  -->
+
+<div>
+    <p>The deploy type will be used.</p>
```
Comment (yungezz) [inline src/main/resources/com/microsoft/jenkins/appservice/WebAppDeploymentRecorder/help-deployType.html]: this is java only, this page will be rendered only for java?

File: pom.xml (hunk 1 above)
Comment (gavinfish) [reply pom.xml]: no, sync with xiangyang, only adding new dependencies should consider azure-commons plugin.

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+public class FileDeployCommand implements ICommand<FileDeployCommand.IWarDeployCommandData> {
+    private String aiDeployInfo;
+    private String aiDeployFailedInfo;
+
+    @Override
+    public void execute(IWarDeployCommandData context) {
+        WebApp app = context.getWebApp();
+        DeployTypeEnum deployType = context.getDeployType();
+
+        FilePath srcDir = context.getJobContext().getWorkspace().child(Util.fixNull(context.getSourceDirectory()));
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: null check on srcDir?

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        try {
+            FilePath[] files = srcDir.list(Util.fixNull(context.getFilePath()));
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: files empty check? if empty list, show some message to user?

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                    }
+                    break;
+                default:
+                    String errorMsg = String.format("Java app does not support Git deployment now. Please choose WAR"
+                            + " or ZIP deployment.", deployType.toString());
+                    throw new IOException(errorMsg);
+            }
+
+            context.setCommandState(CommandState.Success);
+            AzureAppServicePlugin.sendEvent(Constants.AI_WEB_APP, aiDeployInfo,
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: is there way to save user's deploymentType? that will be useful telemetry data

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                    "Run", AppInsightsUtils.hash(context.getJobContext().getRun().getUrl()),
+                    "ResourceGroup", AppInsightsUtils.hash(context.getWebApp().resourceGroupName()),
+                    "WebApp", AppInsightsUtils.hash(context.getWebApp().name()));
+        } catch (IOException e) {
+            context.logError("Fail to deploy war file due to: " + e.getMessage());
```
Comment (yungezz) [inline src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: update error message "war"

Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: done.

File: src/main/java/com/microsoft/jenkins/appservice/util/Constants.java (hunk 3 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/util/Constants.java]: done.

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java (hunk 2 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: done.

File: src/main/resources/com/microsoft/jenkins/appservice/WebAppDeploymentRecorder/help-deployType.html (hunk 4 above)
Comment (gavinfish) [reply src/main/resources/com/microsoft/jenkins/appservice/WebAppDeploymentRecorder/help-deployType.html]: this page will be rendered for all languages, other languages should choose Git deployment.

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java (hunk 5 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: child() method return a new FilePath, cannot be null.

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java (hunk 7 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: done.

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                            context.logStatus(file.getName() + " is not ZIP file. Will skip.");
+                            continue;
+                        }
+                        zipFilesPathList.add(file);
+                    }
+                    if (zipFilesPathList.isEmpty()) {
+                        context.logStatus("Choose ZIP deployment, but there is no zip files in your "
+                                + "setting workspace.");
+                    }
+                    if (zipFilesPathList.size() > 1) {
```
Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: How about making the zip package for user?

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+                    "Run", AppInsightsUtils.hash(context.getJobContext().getRun().getUrl()),
+                    "ResourceGroup", AppInsightsUtils.hash(context.getWebApp().resourceGroupName()),
+                    "WebApp", AppInsightsUtils.hash(context.getWebApp().name()),
+                    "Message", e.getMessage());
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    public interface IWarDeployCommandData extends IBaseCommandData {
```
Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: `IWarDeployCommandData` -> `IFileDeployCommandData`

File: src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentCommandContext.java [hunk 11]
```diff
@@ -147,7 +153,7 @@ public void configure(
             }
         } else if (WebAppUtils.isJavaApp(app)) {
             // For Java application, use WAR deployment
-            builder.withStartCommand(WarDeployCommand.class);
+            builder.withStartCommand(FileDeployCommand.class);
```
Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentCommandContext.java]: Personally speaking, I would rather write two commands `WarDeployCommand` and `ZipDeployCommand` and set different start command here according to deploy type. This makes code cleaner and easier to test. But that's up to you. Just a suggestion.

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java (hunk 9 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: How to make the zip package for user? Package all the files in a folder?

File: src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentCommandContext.java (hunk 11 above)
Comment (gavinfish) [reply src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentCommandContext.java]: Make sense. But won't change since the ZIP&WAR deploy commands are not complicate and they are only for Java apps.

File: src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java (hunk 9 above)
Comment (yungezz) [reply src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: If packaging zip for user, should there be a choice for user to choose what to pack in? what's  the normal user flow without Jenkins? user package all things under build folder directly?

Comment (ArchangelSDY) [reply src/main/java/com/microsoft/jenkins/appservice/commands/FileDeployCommand.java]: In Git deploy, user can specify file name pattern like `**/*.js,**/*.css` to choose what to check in. Zip deploy can be similar.

File: src/test/java/com/microsoft/jenkins/appservice/integration/IntegrationTest.java [hunk 12]
```diff
@@ -89,8 +89,8 @@ To execute the integration tests you need to set the credentials env variables (
             appServicePlanName = TestEnvironment.loadFromEnv("APP_SERVICE_TEST_DEFAULT_APP_SERVICE_PLAN_NAME_PREFIX", "appserviceplan") + "-" + TestEnvironment.GenerateRandomString(16);
 
             SkuDescription sd = new SkuDescription();
-            sd.withTier(TestEnvironment.loadFromEnv("APP_SERVICE_TEST_DEFAULT_PRICING_TIER_SKU", "FREE"));
-            sd.withSize(TestEnvironment.loadFromEnv("APP_SERVICE_TEST_DEFAULT_PRICING_TIER_SIZE", "F1"));
+            sd.withTier(TestEnvironment.loadFromEnv("APP_SERVICE_TEST_DEFAULT_PRICING_TIER_SKU", "SHARED"));
```
Comment (yungezz) [inline src/test/java/com/microsoft/jenkins/appservice/integration/IntegrationTest.java]: why this change?

Comment (gavinfish) [reply src/test/java/com/microsoft/jenkins/appservice/integration/IntegrationTest.java]: free tier only support 10 service plan, I added more test cases which causes total plan counts larger than 10

Comment (gavinfish) [reply src/test/java/com/microsoft/jenkins/appservice/integration/IntegrationTest.java]: Shared tier may charge a little fee, but it is ok since we will destroy the resources as soon as the tests finish
